core/lib/libmbedtls/src/tee_lmd_provider.c

```diff
@@ -767,15 +816,104 @@ void crypto_aes_gcm_final(void *ctx __unused)
 /******************************************************************************
  * Pseudo Random Number Generator
  ******************************************************************************/
-TEE_Result crypto_rng_read(void *buf __unused, size_t blen __unused)
+#if defined(CFG_MBEDTLS_CTR_PRNG)
+static TEE_Result ctr_drbg_read(void *buf, size_t blen)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res = TEE_SUCCESS;
+	int err;
+	mbedtls_ctr_drbg_context ctr_drbg;
+
+	mbedtls_ctr_drbg_init(&ctr_drbg);
+
+	err = mbedtls_ctr_drbg_seed(&ctr_drbg, mbd_rand, NULL, NULL, 0);
+	if (err != 0) {
+		EMSG("failed\n ! mbedtls_ctr_drbg_seed returned 0x%x\n", -err);
+		res = TEE_ERROR_SECURITY;
+		goto exit;
+	}
+
+	err = mbedtls_ctr_drbg_random(&ctr_drbg, buf, blen);
+	if (err != 0) {
+		EMSG("failed!\n");
+		res = TEE_ERROR_BAD_STATE;
+		goto exit;
+	}
+
+exit:
+	mbedtls_ctr_drbg_free(&ctr_drbg);
+	return res;
 }
+#endif
 
-TEE_Result crypto_rng_add_entropy(const uint8_t *inbuf __unused,
-				size_t len __unused)
+#if defined(CFG_MBEDTLS_HMAC_PRNG)
+static TEE_Result hmac_drbg_read(void *buf, size_t blen)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res = TEE_SUCCESS;
+	int err;
+	mbedtls_hmac_drbg_context hmac_drbg;
+	const mbedtls_md_info_t *md_info;
+
+	mbedtls_hmac_drbg_init(&hmac_drbg);
+
+#if defined(MBEDTLS_SHA1_C)
```

Comment:
If both `MBEDTLS_SHA1_C` and `MBEDTLS_SHA256_C` are defined (a very likely scenario) the first call to `mbedtls_md_info_from_type()` doesn't do anything good. Also the code should error (compile time error) out if none of them are defined.

Reply:
Ok, there could to use `if.. else...` to avoid fist problem.
For second problem, at least one of SHA1 and SHA256 must be enabled. So I will define one of them in `mbedtls_config.h` when `CFG_MBEDTLS_HMAC_PRNG` is defined.